src/bin/cql-stress-cassandra-stress/operation/mod.rs

```diff
@@ -16,6 +23,121 @@ use scylla::{
 };
 pub use write::WriteOperationFactory;
 
+use crate::settings::CassandraStressSettings;
+use crate::stats::ShardedStats;
+
+use self::row_generator::RowGenerator;
+
+/// A specific CassandraStress operation.
+#[trait_variant::make(CassandraStressOperation: Send)]
+pub trait LocalCassandraStressOperation: Sync + Send + Sized {
```

Comment:
> Why `LocalCassandraStressOperation` requires `Sync` and `Send`? Shouldn't it require neither?

You are right, these should be moved to non-local version. I forgot to move them after introducing `trait-variant`.

> I'm not sure why a local variant is needed at all.

It actually is not, I just decided to import the `trait-variant` crate and make use of the macro instead of putting the trait bounds on the return future manually (`impl Future<Output = Result<ControlFlow<()>> + Send + Sync>`). I thought that this may look better.

WDYT?

Reply:
> It actually is not, I just decided to import the `trait-variant` crate and make use of the macro instead of putting the trait bounds on the return future manually (`impl Future<Output = Result<ControlFlow<()>> + Send + Sync>`). I thought that this may look better.
> 
> WDYT?

I think it's confusing to use `trait-variant` for that. AFAIK it's primary purpose was to make it possible to quickly generate variants of the trait with different bounds. Using it here implies that the local variant of the trait has some purpose, while it is not used anywhere.

There is only one `async` method in the trait so I think it won't be too bad if we use return position impl in trait (RPITIT) here.

As a side note, why doesn't `CassandraStressOperationFactory::create` have this problem?

Reply:
So I'll remove the `trait-variant` dependency at all for now.

Reply:
@piodul I think that the `Send` bound actually comes from the `async_trait`, not from the `tokio::task::spawn`. The `async_trait` macro on `src/configuration.rs::Operation` expands to:
```rust
fn run<'life0, 'async_trait>(
        &'life0 mut self,
        session: WorkerSession,
    ) -> ::core::pin::Pin<
        Box<dyn ::core::future::Future<Output = Result<()>> + ::core::marker::Send + 'async_trait>,
    >
    where
        'life0: 'async_trait,
        Self: 'async_trait;
```

The error message from the compiler is not clear to me and that's why I assumed that it comes from the `tokio::task::spawn` bound.

For reference - the error message when I remove the `Send` bound from `CassandraStressOperation::execute`:
```
error: future cannot be sent between threads safely
  --> src/bin/cql-stress-cassandra-stress/operation/mod.rs:66:1
   |
66 | make_runnable!(GenericCassandraStressOperation<O: CassandraStressOperation>);
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ future created by async block is not `Send`
   |
   = help: within `{async block@/home/mikolajuzarski/proj/cql-stress/src/configuration.rs:117:100: 125:14}`, the trait `Send` is not implemented for `impl Future<Output = std::result::Result<ControlFlow<()>, anyhow::Error>>`
note: future is not `Send` as it awaits another future which is not `Send`
  --> src/bin/cql-stress-cassandra-stress/operation/mod.rs:80:25
   |
80 |         let op_result = self.cs_operation.execute(row).await;
   |                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ await occurs here on type `impl Future<Output = std::result::Result<ControlFlow<()>, anyhow::Error>>`, which is not `Send`
   = note: required for the cast from `Pin<Box<{async block@/home/mikolajuzarski/proj/cql-stress/src/configuration.rs:117:100: 125:14}>>` to `Pin<Box<dyn Future<Output = std::result::Result<(), anyhow::Error>> + Send>>`
   = note: this error originates in the macro `make_runnable` (in Nightly builds, run with -Z macro-backtrace for more info)

```

Reply:
Makes sense, `async_trait` probably tries to make the returned futures useful in tokio tasks (or, in general, with work stealing executors), so it requires `Send`.